Sources/App/Views/PackageController/PackageShow+Query.swift

```diff
@@ -0,0 +1,116 @@
+import Fluent
+import Foundation
+import Vapor
+
+
+extension PackageShow.Model {
+    static func query(database: Database, packageId: Package.Id) -> EventLoopFuture<Self> {
+        Package.query(on: database)
+            .with(\.$repositories)
+            .with(\.$versions) { $0.with(\.$products) }
+            .filter(\.$id == packageId)
+            .first()
+            .unwrap(or: Abort(.notFound))
+            .map { p -> Self? in
+                // we consider certain attributes as essential and return nil (raising .notFound)
+                guard let title = p.name else { return nil }
+                return Self.init(title: title,
+                                 url: p.url,
+                                 license: p.repository?.license ?? .none,
+                                 summary: p.repository?.summary ?? "–",
+                                 authors: [],      // TODO: fill in
+                                 history: nil,     // TODO: fill in
+                                 activity: nil,    // TODO: fill in
+                                 products: p.productCounts,
+                                 releases: p.releaseInfo,
+                                 languagePlatforms: .init(
+                                    stable: .init(
+                                        link: .init(name: "stable", url: "stable"),  // TODO: fill in
+                                        swiftVersions: [],                           // TODO: fill in
+                                        platforms: []),                              // TODO: fill in
+                                    beta: .init(
+                                        link: .init(name: "beta", url: "beta"),      // TODO: fill in
+                                        swiftVersions: [],                           // TODO: fill in
+                                        platforms: []),                              // TODO: fill in
+                                    latest: .init(
+                                        link: .init(name: "latest", url: "latest"),  // TODO: fill in
+                                        swiftVersions: [],                           // TODO: fill in
+                                        platforms: [])))                             // TODO: fill in
+            }
+            .unwrap(or: Abort(.notFound))
+    }
+}
+
+
+extension Package {
+    var defaultVersion: Version? {
+        guard
+            let versions = $versions.value,
+            let repositories = $repositories.value,
```

Comment:
NB: after discussing with Tanner Nelson in Discord I've arrived at this way to safely access relationship data: `model.$relationship.value` will return a double optional if the data is not loaded and a single optional if the data is loaded but nil (in case of optional relationships).

That means we can now safely use these helpers. Safely in the sense that we won't see crashes. The interface is still less than ideal, because inspecting a double optional to detect non-fulfilled relationships isn't great.

I'll probably move some of these extension over to `Package.swift` now that they're ok to use but I want to see how wiring everything up goes first.